hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/AvroConversionUtils.scala

```diff
@@ -41,8 +107,8 @@ object AvroConversionUtils {
         else {
           val schema = new Schema.Parser().parse(schemaStr)
           val dataType = convertAvroSchemaToStructType(schema)
-          val convertor = AvroConversionHelper.createConverterToRow(schema, dataType)
-          records.map { x => convertor(x).asInstanceOf[Row] }
+          val converter = createConverterToRow(schema, dataType)
```

Comment:
These are different APIs: one returns `Row` while other returns (Spark's internal) `InternalRow` allowing to avoid penalty of deserialization from one into another.

The plan is to gradually weed out (where possible) replace `Row` converter w/ `InternalRow` one

Reply:
@yihua you brought up legitimate concerns, there are a few considerations here

 - First of all, we already depend on `InternalRow` quite a bit (we even have out own `HoodieInternalRow` extension)
 - `InternalRow` is a core component of Spark that is unlikely to change substantially (as that would mean that quite a bit of Spark will have to be re-written to accommodate for any substantial changes to it, which, again, i don't think are likely)

At the same time avoiding `InternalRow` > `Row` conversion has some considerable performance advantages  , and it's exactly how Spark operates internally: all of its internal `Plan`s, expressions, operators, operate on `InternalRow` and defer such deserialization (`InternalRow` to `Row`) only to cases when you dereference it to `RDD[Row]`.

All in all, for us to be able to compete w/ Delta on performance we will have to pull the same tricks they are using and make sure our overhead is as lean as possible

Reply:
>we even have out own HoodieInternalRow extension
This is a different code path altogether. 

but I think its okay to depend on InternalRow. Spark's undocumented APIs are a pain regardless

Reply:
To be clear, we already have a row writer path, which is exactly on par with any other framework's parquet writing. On the upsert path, where this code is used, the bottleneck is indexing and we already have significant speedups due to the rich set of indexing schemes.That said, using internalrow is ok

Reply:
Sg.  Based on the info provided, in this case, looks like we can depend on `InternalRow` for now, for performance reasons.  For Row writer path, I think if we can leverage Spark's internal optimization with `InternalRow` that'd be good, without using `InternalRow` directly in Hudi code.

Reply:
@yihua i think even in row writing path, we still need to use `InternalRow` quite a bit, to achieve some data manipulation. Sorry i made the discussion way out of scope from here. iceberg made a similar extension to InternalRow for writing https://github.com/apache/iceberg/blob/24896d84c5/spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java
cc @vinothchandar @alexeykudinkin @nsivabalan